### DISCUSSION TITLE
Using scripts in corporate environment

### DISCUSSION THREAD

**Discussion by jiramares** [VirtusLab/scala-cli#3371] [Scripting]
Hi guys,

I have problem with using `*.sc` files with `scala-cli` in corporate environment. First there is huge problem of coursier with repositories accessed via SSL with self signed certificates (it is nearly unusable). We cannot integrate our certificate into coursier anyway only via command line.

But to be humble I doesn't want to scala-cli to resolve some dependencies everytime. I would like to set up the environment (whether manually or automatically via coursier and its machinery doesn't matter) and then use it.

Next comes building server that is not re-entrant so when run more scripts in one command (via pipe in shell) or when run concurrently from more shells we get sometimes strange behavior (as described in #2034). If I turn off using compile server (`--server=false`) everything works but it is terribly slow.

So next iteration, make from `*.sc` file runnable jar (via `package` sub-command). Yes, that is great and works ... but we are back with the coursier problem and self-signed certificate (the repository reference is hard coded into the jar someway). I can for each script make self-contained huge jar (I doesn't want to) or I have small jar with coursier inside and the problem with self-singed certificate.

Is it possible to create runnable jar with external dependencies (copies to some directory and references in manifest inside jar) without coursier inside that can be copies (with its external dependencies) and works?

Thanks

PS. At the moment the cleanest solution  I see is make `sbt` project, transform `*.sc` files into `*.scala` files with main method and create runnable jar on my own. Than copy the jar and its dependencies and use it.

Jirka

**Comment by Gedochao** [comment]
> First there is huge problem of coursier with repositories accessed via SSL with self signed certificates (it is nearly unusable). We cannot integrate our certificate into coursier anyway only via command line.

@jiramares can you elaborate on the `coursier` certificate problem? did you raise an issue in the `coursier` backlog? Maybe we can just figure out how to make it work for you.

> Next comes building server that is not re-entrant so when run more scripts in one command (via pipe in shell) or when run concurrently from more shells we get sometimes strange behavior (as described in https://github.com/VirtusLab/scala-cli/issues/2034). If I turn off using compile server (--server=false) everything works but it is terribly slow.

Let's track this in #2034... it seems like a bug, although I'm not sure what is causing it.

> So next iteration, make from *.sc file runnable jar (via package sub-command). Yes, that is great and works ... but we are back with the coursier problem and self-signed certificate (the repository reference is hard coded into the jar someway). I can for each script make self-contained huge jar (I doesn't want to) or I have small jar with coursier inside and the problem with self-singed certificate.

Once again, let's try to figure out how to make `coursier` work for you. I'm not sure what exactly breaks.

> Is it possible to create runnable jar with external dependencies (copies to some directory and references in manifest inside jar) without coursier inside that can be copies (with its external dependencies) and works?

So... I'm not sure I understand correctly, but you'd like to be able to create a self-executable JAR which points to locally downloaded dependencies, rather than rely on `coursier`?That in and by itself sounds like a feature request to me (and one that I don't think is likely we would add, although feel free to raise a ticket, we can discuss it separately)... 

Some (not necessarily pretty) workarounds that come to mind:
- prefill your [`coursier` cache](https://get-coursier.io/docs/cache) and run with `--offline`
- package into a `--library` JAR and provide the classpath when running the JAR
  - You could make a short script for running the JAR by hand, something like:
  ```bash
  java -cp YourJar.jar:AnotherJar.jar* your.main.class.Main
  ```
  - or just use Scala CLI/`coursier launch` to run the Jar with the classpath (the standard java runner doesn't support `--jar` with `-cp`)
  ```bash
   scala-cli YourJar.jar -cp yourJarsClasspath -M your.main.class.Main
  ```
  - you can get the classpath of a Scala CLI project with `--print-classpath`
  ```bash
  scala-cli compile . --print-classpath
  ```

**Comment by jiramares** [comment]
I am sorry for so late reaction.

First I would like to resolve dependencies via our local artifact repository and the server running this repository is not reachable from DMZ (there are our scripts deployed too). Manual copying all files in right directories to mimic downloading via `cousier` is at least time-consuming.

I chose the workaround using ssh remote port forwarding. From computer that can reach DMZ via ssh a can communicate with artifact repository connect via ssh to server and port forward server local port to artifact repository. In `/etc/hosts` define alias for localhost with name of the artifact repository server. Now I can make artifact repository visible in my DMZ server for period of time that I control.

Second comes `coursier` problems:

1) If you set up repositories using `scala-cli config` this setting affects `sbt` too so we have to use environment variable `COURSIER_REPOSITORIES` that affects `cs` command for a change
2) this settings affect also the `cs` command and included `coursier` for resolving artifacts in built jars from scripts (using `scala-cli package` command) and in these commands I can't specify `-Djavax.net.ssl.trustStore` and therefore resolution is not working

These problems has also workarounds:

1) when using `cs` command reset the COURSIER_REPOSITORIES 
2) when run build `scala-cli` scripts resolve dependencies manually by running `scala-cli -Djavax.net.ssl.trustStore=my/trust/store/cacerts some-script-wtith-right-dependencies.sc`

But I don't see that as easy

Therefore I asked to make another way to pre-package everything I need to run the script (ideally also built because concurrent build problem) and deploy it (not include the `coursier` to resolve anything in runtime).
